File: ex07/countPennies.cpp
#include <opencv2/opencv.hpp>
#include <vector>
#include <algorithm>
#include <cmath>
#include <iostream>

struct Circle {
    int x, y, radius;
};

std::vector<Circle> non_max_suppression(const std::vector<Circle>& circles, float overlap_threshold = 0.5f) {
    if (circles.empty()) {
        return {};
    }

    // Sort circles by radius (largest first)
    std::vector<Circle> sorted_circles = circles;
    std::sort(sorted_circles.begin(), sorted_circles.end(), [](const Circle& a, const Circle& b) {
        return a.radius > b.radius;
    });

    std::vector<Circle> suppressed;

    while (!sorted_circles.empty()) {
        // Take the largest circle
        Circle current = sorted_circles.front();
        suppressed.push_back(current);
        sorted_circles.erase(sorted_circles.begin());

        // Calculate overlap with remaining circles
        std::vector<Circle> to_keep;
        for (const auto& circle : sorted_circles) {
            // Calculate distance between centers
            float dx = current.x - circle.x;
            float dy = current.y - circle.y;
            float distance = std::sqrt(dx * dx + dy * dy);

            // Calculate overlap ratio (intersection over union)
            if (distance < (current.radius + circle.radius)) {
                float overlap_ratio = (current.radius * current.radius) / (circle.radius * circle.radius);
                if (overlap_ratio <= overlap_threshold) {
                    to_keep.push_back(circle);
                }
            } else {
                to_keep.push_back(circle);
            }
        }

        sorted_circles = to_keep;
    }

    return suppressed;
}

int detect_circles(const std::string& image_path, bool visualize = false) {
    // Load the image
    cv::Mat image = cv::imread(image_path);
    if (image.empty()) {
        std::cerr << "Error: Image not found." << std::endl;
        return 0;
    }

    // Convert to grayscale
    cv::Mat gray;
    cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);

    // Apply Gaussian blur to reduce noise
    cv::Mat blurred;
    cv::GaussianBlur(gray, blurred, cv::Size(9, 9), 2);

    // Apply Canny edge detection
    cv::Mat edges;
    cv::Canny(blurred, edges, 50, 150);

    // Create a binary version of the image
    cv::Mat binary;
    cv::threshold(edges, binary, 50, 255, cv::THRESH_BINARY);

    // Detect circles using Hough Circle Transform on the binarized image
    std::vector<cv::Vec3f> circles;
    cv::HoughCircles(binary, circles, cv::HOUGH_GRADIENT, 1.2, 30, 50, 30, 10, 100);

    if (!circles.empty()) {
        // Convert to Circle struct
        std::vector<Circle> detected_circles;
        for (const auto& circle : circles) {
            detected_circles.push_back({static_cast<int>(circle[0]), static_cast<int>(circle[1]), static_cast<int>(circle[2])});
        }

        // Apply Non-Maximum Suppression (NMS)
        std::vector<Circle> filtered_circles = non_max_suppression(detected_circles, 0.5f);

        if (!filtered_circles.empty()) {
            int num_circles = filtered_circles.size();

            if (visualize) {
                // Draw the detected circles on the original image
                for (const auto& circle : filtered_circles) {
                    // Draw the outer circle
                    cv::circle(image, cv::Point(circle.x, circle.y), circle.radius, cv::Scalar(0, 255, 0), 2);
                    // Draw the center of the circle
                    cv::circle(image, cv::Point(circle.x, circle.y), 2, cv::Scalar(0, 0, 255), 3);
                }

                // Show the output image with detected circles
                cv::imshow("Detected Circles", image);
                cv::waitKey(0);
                cv::destroyAllWindows();
            }

            return num_circles;
        } else {
            std::cout << "No circular objects detected after NMS." << std::endl;
            return 0;
        }
    } else {
        std::cout << "No circular objects detected." << std::endl;
        return 0;
    }
}

int main(int argc, char** argv) {
    if (argc < 2) {
        std::cerr << "Usage: " << argv[0] << " -i <image_path> [-v]" << std::endl;
        return -1;
    }

    std::string image_path;
    bool visualize = false;

    for (int i = 1; i < argc; ++i) {
        if (std::string(argv[i]) == "-i" && i + 1 < argc) {
            image_path = argv[i + 1];
        } else if (std::string(argv[i]) == "-v") {
            visualize = true;
        }
    }

    if (image_path.empty()) {
        std::cerr << "Error: Image path not provided." << std::endl;
        return -1;
    }

    // Detect circles in the image
    int num_circles = detect_circles(image_path, visualize);

    // Print the number of detected circles
    std::cout << "Number of circular objects detected: " << num_circles << std::endl;

    return 0;
}

File: ex07/countPennies.py
import argparse
import cv2
import numpy as np

def non_max_suppression(circles, overlap_threshold=0.5):
    """
    Apply Non-Maximum Suppression (NMS) to filter out overlapping circles.

    Parameters:
        circles (np.array): Array of detected circles in the format (x, y, radius).
        overlap_threshold (float): Threshold for overlapping area ratio to suppress a circle.

    Returns:
        np.array: Filtered circles after NMS.
    """
    if circles is None or len(circles) == 0:
        return np.array([])

    # Convert circles to a list of (x, y, radius) tuples
    circles = np.uint16(np.around(circles))
    circles = circles[0, :]  # Remove the extra dimension
    circles = sorted(circles, key=lambda x: x[2], reverse=True)  # Sort by radius (largest first)

    suppressed = []

    while len(circles) > 0:
        # Take the largest circle
        current = circles[0]
        suppressed.append(current)
        circles = circles[1:]

        # Calculate overlap with remaining circles
        to_remove = []
        for i, circle in enumerate(circles):
            # Calculate distance between centers
            dx = current[0] - circle[0]
            dy = current[1] - circle[1]
            distance = np.sqrt(dx**2 + dy**2)

            # Calculate overlap ratio (intersection over union)
            if distance < (current[2] + circle[2]):
                # Circles overlap
                overlap_ratio = (current[2] ** 2) / (circle[2] ** 2)
                if overlap_ratio > overlap_threshold:
                    to_remove.append(i)

        # Remove suppressed circles
        circles = [circle for i, circle in enumerate(circles) if i not in to_remove]

    return np.array([suppressed])


def detect_circles(image_path, visualize=False):
    """
    Detects circular objects in an image using Hough Circle Transform and applies NMS.

    Parameters:
        image_path (str): Path to the input image.
        visualize (bool): Whether to visualize the detected circles on the output image.

    Returns:
        int: Number of circular objects detected.
    """
    # Load the image
    image = cv2.imread(image_path)
    if image is None:
        print("Error: Image not found.")
        return 0

    # Convert to grayscale
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    # Apply Gaussian blur to reduce noise
    blurred = cv2.GaussianBlur(gray, (9, 9), 2)
    
    edges = cv2.Canny(blurred, 50, 150)

    # Detect circles using Hough Circle Transform
    circles = cv2.HoughCircles(
        edges,
        cv2.HOUGH_GRADIENT,
        dp=1.2,  # Inverse ratio of the accumulator resolution
        minDist=30,  # Minimum distance between detected centers
        param1=50,  # Upper threshold for edge detection
        param2=30,  # Threshold for center detection
        minRadius=10,  # Minimum radius of circles to detect
        maxRadius=500   # Maximum radius of circles to detect
    )

    if circles is not None:
        # Apply Non-Maximum Suppression (NMS)
        filtered_circles = non_max_suppression(circles, overlap_threshold=0.5)

        if len(filtered_circles) > 0:
            num_circles = len(filtered_circles[0, :])

            if visualize:
                # Draw the detected circles on the original image
                for circle in filtered_circles[0, :]:
                    x, y, radius = circle
                    # Draw the outer circle
                    cv2.circle(image, (x, y), radius, (0, 255, 0), 2)
                    # Draw the center of the circle
                    cv2.circle(image, (x, y), 2, (0, 0, 255), 3)

                # Show the output image with detected circles
                cv2.imshow("Detected Circles", image)
                cv2.waitKey(0)
                cv2.destroyAllWindows()

            return num_circles
        else:
            print("No circular objects detected after NMS.")
            return 0
    else:
        print("No circular objects detected.")
        return 0


if __name__ == "__main__":
    # Set up argument parser
    parser = argparse.ArgumentParser(description="Detect circular objects in an image with NMS.")
    parser.add_argument("-i", "--image", required=True, help="Path to the input image.")
    parser.add_argument("-v", "--visualize", action="store_true", help="Visualize the detected circles on the output image.")

    # Parse arguments
    args = parser.parse_args()

    # Detect circles in the image
    num_circles = detect_circles(args.image, visualize=args.visualize)

    # Print the number of detected circles
    print(f"Number of circular objects detected: {num_circles}")